presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,189 @@
+        }
+        else {
+            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            outputFunction = LONG_DECIMAL_OUTPUT_FUNCTION;
+        }
+        inputFunction = inputFunction.bindTo(type);
+        outputFunction = outputFunction.bindTo(type);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type, inputTypes),
+                createInputParameterMetadata(type),
+                inputFunction,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                outputFunction,
+                stateInterface,
+                stateSerializer,
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(BigInteger.valueOf(type.getLong(block, position)), state);
+    }
+
+    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(unscaledValueToBigInteger(type.getSlice(block, position)), state);
+    }
+
+    private static void accumulateValueInState(BigInteger value, BigIntegerAndLongState state)
+    {
+        initializeIfNeeded(state);
+        state.setBigInteger(state.getBigInteger().add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(otherState.getBigInteger());
+        }
+        else {
+            state.setBigInteger(state.getBigInteger().add(otherState.getBigInteger()));
+        }
+    }
+
+    public static void outputShortDecimal(ShortDecimalType type, BigIntegerAndLongState state, BlockBuilder out)
+    {
+        if (state.getLong() == 0) {
+            out.appendNull();
+        }
+        else {
+            type.writeLong(out, average(state, type).unscaledValue().longValue());
```

Comment:
done, used `BigInteger.longValueExact()` which throws an `ArithmeticException` if it doesn't fit